File: scripts/fetch_web.py
import os
import libsql_experimental as libsql
import json
from datetime import datetime
import sys
import requests
from bs4 import BeautifulSoup

def create_turso_tables(conn):
    """
    Create the necessary tables in Turso database
    """
    print("Creating tables in Turso database...")
    
    # Create content table
    conn.execute('''
    CREATE TABLE IF NOT EXISTS content (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        url TEXT UNIQUE,
        content_type TEXT,
        title TEXT,
        created_at TEXT,
        consumed_at TEXT,
        last_updated_at TEXT,
        scraped_at TEXT,
        is_scraped INTEGER DEFAULT 0
    )
    ''')
    
    # Create metadata table
    conn.execute('''
    CREATE TABLE IF NOT EXISTS metadata (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        content_id INTEGER,
        og_title TEXT,
        og_description TEXT,
        og_image TEXT,
        keywords TEXT,
        FOREIGN KEY (content_id) REFERENCES content(id)
    )
    ''')
    
    # Create web table
    conn.execute('''
    CREATE TABLE IF NOT EXISTS web (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        content_id INTEGER,
        url TEXT,
        published_at TEXT,
        full_content TEXT,
        FOREIGN KEY (content_id) REFERENCES content(id)
    )
    ''')
    
    # Create YouTube tables - for future use
    conn.execute('''
    CREATE TABLE IF NOT EXISTS youtube (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        content_id INTEGER,
        url TEXT,
        video_id TEXT,
        channel_name TEXT,
        description TEXT,
        duration TEXT,
        FOREIGN KEY (content_id) REFERENCES content(id)
    )
    ''')
    
    conn.execute('''
    CREATE TABLE IF NOT EXISTS transcript (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        youtube_id INTEGER,
        full_text TEXT,
        language TEXT,
        duration REAL,
        fetched_at TEXT,
        FOREIGN KEY (youtube_id) REFERENCES youtube(id)
    )
    ''')
    
    conn.execute('''
    CREATE TABLE IF NOT EXISTS transcript_segments (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        transcript_id INTEGER,
        start_time TEXT,
        end_time TEXT,
        text TEXT,
        FOREIGN KEY (transcript_id) REFERENCES transcript(id)
    )
    ''')
    
    # Create sync_history table
    conn.execute('''
    CREATE TABLE IF NOT EXISTS sync_history (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        sync_time TEXT,
        entries_added INTEGER,
        entries_updated INTEGER,
        entries_scraped INTEGER,
        scrape_errors INTEGER,
        sync_type TEXT
    )
    ''')
    
    print("Tables created successfully.")

def fetch_web_content(api_endpoint):
    """
    Fetch web content from the API
    """
    print(f"Fetching web content from {api_endpoint}...")
    try:
        response = requests.get(api_endpoint)
        response.raise_for_status()
        
        # Print the first 100 characters to debug
        print(f"API Response (first 100 chars): {response.text[:100]}...")
        
        # Try to parse as JSON
        try:
            return response.json()
        except json.JSONDecodeError as e:
            print(f"Failed to decode JSON: {e}")
            # If it's a string, try to parse URLs directly
            lines = response.text.strip().split('\n')
            return [{"url": line.strip()} for line in lines if line.strip()]
    except Exception as e:
        print(f"Error fetching content: {str(e)}")
        return []

def process_and_store_content(content_data, turso_conn):
    """
    Process fetched content and store in Turso
    """
    entries_added = 0
    scrape_errors = 0
    
    # Handle different data formats
    if not content_data:
        print("No content data to process")
        return 0, 0
        
    # Check if content_data is a list or dict
    if isinstance(content_data, str):
        # Handle string data (possibly a URL list)
        try:
            # Try to parse as JSON string first
            content_data = json.loads(content_data)
        except json.JSONDecodeError:
            # Split by lines if it's a text list of URLs
            content_data = [{"url": line.strip()} for line in content_data.split('\n') if line.strip()]
    
    # Ensure we have a list to iterate
    if not isinstance(content_data, list):
        content_data = [content_data]
    
    for item in content_data:
        url = None
        try:
            # Handle different item formats
            if isinstance(item, dict):
                url = item.get('url')
            elif isinstance(item, str):
                url = item
                
            if not url:
                continue
                
            print(f"Processing URL: {url}")
                
            # Check if URL already exists
            existing = turso_conn.execute("SELECT id FROM content WHERE url = ?", (url,)).fetchone()
            if existing:
                print(f"URL already exists: {url}")
                continue
                
            # Fetch the content if needed
            try:
                content = ""
                title = ""
                if isinstance(item, str) or (isinstance(item, dict) and 'content' not in item):
                    print(f"Fetching content for {url}")
                    page_response = requests.get(url, timeout=10)
                    if page_response.status_code == 200:
                        soup = BeautifulSoup(page_response.text, 'html.parser')
                        title = soup.title.string if soup.title else ""
                        content = page_response.text
                else:
                    title = item.get('title', '')
                    content = item.get('content', '')
            except Exception as e:
                print(f"Error fetching content for {url}: {e}")
                content = ""
                title = ""
            
            # Add new content entry
            current_time = datetime.now().isoformat()
            turso_conn.execute('''
            INSERT INTO content (url, content_type, title, created_at)
            VALUES (?, ?, ?, ?)
            ''', (url, 'web', title, current_time))
            
            # Get the content_id
            content_id = turso_conn.execute("SELECT id FROM content WHERE url = ?", (url,)).fetchone()[0]
            
            # Add to web table
            turso_conn.execute('''
            INSERT INTO web (content_id, url, full_content)
            VALUES (?, ?, ?)
            ''', (content_id, url, content))
            
            # Add metadata if available
            meta = {}
            if isinstance(item, dict) and 'metadata' in item:
                meta = item['metadata']
            
            turso_conn.execute('''
            INSERT INTO metadata (content_id, og_title, og_description, og_image, keywords)
            VALUES (?, ?, ?, ?, ?)
            ''', (
                content_id,
                meta.get('og_title', '') if isinstance(meta, dict) else '',
                meta.get('og_description', '') if isinstance(meta, dict) else '',
                meta.get('og_image', '') if isinstance(meta, dict) else '',
                json.dumps(meta.get('keywords', [])) if isinstance(meta, dict) else '[]'
            ))
            
            entries_added += 1
            print(f"Added entry for {url}")
            
        except Exception as e:
            print(f"Error processing item {url if url else 'unknown'}: {str(e)}")
            scrape_errors += 1
    
    # Add to sync history
    current_time = datetime.now().isoformat()
    turso_conn.execute('''
    INSERT INTO sync_history (
        sync_time, entries_added, entries_updated, entries_scraped, scrape_errors, sync_type
    ) VALUES (?, ?, ?, ?, ?, ?)
    ''', (
        current_time,
        entries_added,
        0,
        0,
        scrape_errors,
        "api_fetch"
    ))
    
    return entries_added, scrape_errors

def main():
    # Check environment variables or use defaults
    API_ENDPOINT = os.environ.get("API_ENDPOINT", "https://open-source-content.xyz/v1/web")
    TURSO_URL = os.environ.get("TURSO_URL", "libsql://context-amantulsyan35.aws-us-east-1.turso.io")
    TURSO_AUTH_TOKEN = os.environ.get("TURSO_AUTH_TOKEN")
    
    # Validate that we have the Turso auth token
    if not TURSO_AUTH_TOKEN:
        print("Error: TURSO_AUTH_TOKEN environment variable is required")
        sys.exit(1)
    
    # Connect to Turso 
    print(f"Connecting to Turso database: {TURSO_URL}")
    # The local.db is just a local cache file
    turso_conn = None
    
    try:
        turso_conn = libsql.connect("local.db", sync_url=TURSO_URL, auth_token=TURSO_AUTH_TOKEN)
        
        # Create Turso tables if they don't exist
        create_turso_tables(turso_conn)
        
        # Fetch web content
        content_data = fetch_web_content(API_ENDPOINT)
        
        # Process and store content
        if content_data:
            entries_added, scrape_errors = process_and_store_content(content_data, turso_conn)
            print(f"Added {entries_added} new entries with {scrape_errors} errors")
        else:
            print("No content fetched from API")
        
        # Sync changes to Turso
        print("Syncing changes to Turso...")
        turso_conn.sync()
        print("Sync completed successfully.")
        
        # Validate by counting rows
        content_count = turso_conn.execute("SELECT COUNT(*) FROM content").fetchone()[0]
        print(f"Current content entries in database: {content_count}")
            
    except Exception as e:
        print(f"Error during execution: {str(e)}")
        if turso_conn:
            try:
                turso_conn.sync()
            except:
                pass
        sys.exit(1)
    finally:
        # Close connection - safely
        if turso_conn:
            # libsql Connection might not have close() method
            try:
                turso_conn.sync()  # Make sure changes are synced
            except:
                pass
        print("Database operation completed.")

if __name__ == "__main__":
    main()

File: scripts/fetch_web.ts
import { load } from "cheerio";
import { Pool } from "pg";
import OpenAI from "openai";
import { SQL } from "../sql-queries";
import { OpenSourceEntry } from "../types";

// Initialize OpenAI client
const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

// Initialize PostgreSQL pool
const pool = new Pool({
  host: process.env.POSTGRES_HOST,
  port: 5432,
  database: process.env.POSTGRES_DB,
  user: process.env.POSTGRES_USER,
  password: process.env.POSTGRES_PASSWORD,
  ssl: {
    rejectUnauthorized: false,
  },
});

// Types
interface ExtractedContent {
  title: string;
  url: string;
  publishedAt: string | null;
  fullContent: string;
  metaData: {
    ogTitle: string;
    ogDescription: string;
    ogImage: string;
    keywords: string;
  };
}

interface ContentWithEmbedding extends ExtractedContent {
  embedding: number[];
}

function formatEmbeddingForPostgres(embedding: number[]): string {
  // Format the array for PostgreSQL vector type - should start with [
  return `[${embedding.join(",")}]`;
}

async function extractPageContent(link: string): Promise<ExtractedContent> {
  try {
    console.log(`Fetching content for ${link}`);

    // Set up AbortController for timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 10000);

    try {
      // Fetch the HTML content from the link with headers
      const response = await fetch(link, {
        headers: {
          "User-Agent":
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
        },
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new Error(
          `Failed to fetch ${link}: Status code ${response.status}`
        );
      }

      const html = await response.text();

      // Load the HTML content into cheerio
      const $ = load(html);

      // Extract the page title
      const title =
        $("title").text().trim() || $("h1").first().text().trim() || link;

      // Collect metadata
      const metaData = {
        ogTitle: $('meta[property="og:title"]').attr("content") || "",
        ogDescription:
          $('meta[property="og:description"]').attr("content") || "",
        ogImage: $('meta[property="og:image"]').attr("content") || "",
        keywords: $('meta[name="keywords"]').attr("content") || "",
      };

      let publishedAt: any;

      // Try to find date in common meta tags
      let publishedTimeRaw =
        $('meta[property="article:published_time"]').attr("content") ||
        $('meta[name="date"]').attr("content") ||
        $("time").attr("datetime") ||
        "";

      if (publishedTimeRaw) {
        publishedAt = publishedTimeRaw;
      }

      // Extract full page content, focusing on meaningful text elements
      // Remove script, style, and other non-content elements
      $(
        "script, style, noscript, iframe, img, svg, path, head, nav, footer, aside"
      ).remove();

      // Get all text content, preserving some structure
      let fullContent = "";

      // Get heading elements with their hierarchy
      $("h1, h2, h3, h4, h5, h6").each((_, element) => {
        const tagName = element.tagName?.toLowerCase?.();
        const level = parseInt(tagName?.substring(1) || "0");
        const prefix = "#".repeat(level || 1) + " ";
        const headingText = $(element).text().trim();
        fullContent += prefix + headingText + "\n\n";
      });

      // Extract paragraphs
      $(
        "p, article, section, div, main, span, li, td, th, blockquote, pre, code, figcaption"
      ).each((_, element) => {
        const text = $(element).text().trim();
        if (text && text.length > 10) {
          // Skip very short elements that might just be styling
          fullContent += text + "\n\n";
        }
      });

      // Remove excessive whitespace
      fullContent = fullContent
        .replace(/(\n\s*){3,}/g, "\n\n") // Replace 3+ line breaks with 2
        .replace(/\s{2,}/g, " ") // Replace multiple spaces with a single space
        .trim();

      return {
        title,
        url: link,
        publishedAt,
        fullContent,
        metaData,
      };
    } catch (fetchError: any) {
      clearTimeout(timeoutId);
      if (fetchError.name === "AbortError") {
        throw new Error(`Request timeout for ${link}`);
      }
      throw fetchError;
    }
  } catch (error) {
    console.error(`Error scraping ${link}:`, error);

    // Return default values in case of error
    return {
      title: link, // Using the URL as the title for failed pages
      url: link,
      publishedAt: null,
      fullContent: `Failed to retrieve content from ${link}.`,
      metaData: {
        ogTitle: "",
        ogDescription: "",
        ogImage: "",
        keywords: "",
      },
    };
  }
}

async function generateEmbedding(text: string): Promise<number[]> {
  try {
    // Truncate text if it's too long (OpenAI has token limits)
    const truncatedText = text.substring(0, 8000);

    // Generate embedding using OpenAI API
    const response = await openai.embeddings.create({
      model: "text-embedding-ada-002",
      input: truncatedText,
    });

    return response.data[0].embedding;
  } catch (error) {
    console.error("Error generating embedding:", error);
    throw error;
  }
}

async function storeContentInDatabase(
  content: ContentWithEmbedding,
  consumedAt: string
): Promise<boolean> {
  const client = await pool.connect();

  try {
    const currentTime = new Date().toISOString();
    const consumedTimestamp = consumedAt || null;

    // Begin transaction
    await client.query("BEGIN");

    // Check if URL already exists
    const existingResult = await client.query(SQL.CHECK_URL_EXISTS, [
      content.url,
    ]);

    if (existingResult.rows.length > 0) {
      console.log(`URL already exists: ${content.url} - skipping`);
      await client.query("COMMIT");
      return false;
    }

    // Insert into content table
    const contentResult = await client.query(
      `INSERT INTO content
      (url, content_type, title, created_at, consumed_at, scraped_at, is_scraped)
      VALUES ($1, $2, $3, $4, $5, $6, $7)
      RETURNING id`,
      [
        content.url,
        "web",
        content.title,
        currentTime,
        consumedTimestamp, // Store the createdAt as consumedAt
        currentTime,
        true,
      ]
    );

    const contentId = contentResult.rows[0].id;

    // Format the embedding properly for PostgreSQL
    const formattedEmbedding = formatEmbeddingForPostgres(content.embedding);

    // Insert into web table with properly formatted embedding
    await client.query(SQL.INSERT_WEB, [
      contentId,
      content.url,
      content.publishedAt,
      content.fullContent,
      formattedEmbedding, // Use the formatted embedding
    ]);

    // Insert into metadata table
    await client.query(SQL.INSERT_METADATA, [
      contentId,
      content.metaData.ogTitle,
      content.metaData.ogDescription,
      content.metaData.ogImage,
      content.metaData.keywords,
    ]);

    // Commit transaction
    await client.query("COMMIT");

    console.log(`Successfully added entry for ${content.url}`);
    return true;
  } catch (error) {
    // Rollback in case of error
    await client.query("ROLLBACK");
    console.error(`Database error for ${content.url}:`, error);
    return false;
  } finally {
    client.release();
  }
}

async function processURLs(
  entries: OpenSourceEntry[],
  limit: number = 10
): Promise<{ added: number; errors: number; skipped: number }> {
  // Limit to first N entries if needed
  const entriesToProcess = entries.slice(0, limit);
  console.log(`Processing ${entriesToProcess.length} URLs (limit: ${limit})`);

  let addedCount = 0;
  let errorCount = 0;
  let skippedCount = 0;

  try {
    // Process each entry
    for (const entry of entriesToProcess) {
      try {
        const url = entry.url;

        // Check if URL already exists
        const client = await pool.connect();
        const existingResult = await client.query(SQL.CHECK_URL_EXISTS, [url]);
        client.release();

        if (existingResult.rows.length > 0) {
          console.log(`URL already exists: ${url}`);
          skippedCount++;
          continue;
        }

        // 1. Extract content
        const content = await extractPageContent(url);

        // 2. Generate embedding from content
        const embedding = await generateEmbedding(content.fullContent);

        // 3. Store in database with embedding
        const contentWithEmbedding: ContentWithEmbedding = {
          ...content,
          embedding,
        };

        // Pass the createdAt timestamp to be used as consumedAt
        const success = await storeContentInDatabase(
          contentWithEmbedding,
          entry.createdAt
        );

        if (success) {
          addedCount++;
        } else {
          skippedCount++;
        }
      } catch (error) {
        console.error(`Error processing ${entry.url}:`, error);
        errorCount++;
      }
    }

    // Add entry to sync_history
    const client = await pool.connect();
    const currentTime = new Date().toISOString();
    await client.query(SQL.INSERT_SYNC_HISTORY, [
      currentTime,
      addedCount,
      0,
      addedCount,
      errorCount,
      "api_fetch",
    ]);

    // Get the count of content entries
    const countResult = await client.query(SQL.GET_CONTENT_COUNT);
    const contentCount = countResult.rows[0].count;
    console.log(`Current content entries in database: ${contentCount}`);
    client.release();
  } catch (error) {
    console.error("Error during execution:", error);
    throw error;
  }

  return { added: addedCount, errors: errorCount, skipped: skippedCount };
}

// Main function to fetch and process content
async function main() {
  try {
    const apiEndpoint =
      process.env.API_ENDPOINT || "https://open-source-content.xyz/v1/web";
    console.log(`Fetching web content from ${apiEndpoint}...`);

    // Fetch the list of entries from the API
    const response = await fetch(apiEndpoint);
    const data = await response.json();

    // Extract entries from the data
    let entries: OpenSourceEntry[] = [];
    if (data && data.data && Array.isArray(data.data)) {
      // Filter to ensure we have valid entries (with url field)
      entries = data.data.filter(
        (entry: any) =>
          typeof entry === "object" &&
          entry !== null &&
          typeof entry.url === "string"
      );

      // Filter out YouTube URLs
      entries = entries.filter(
        (entry: OpenSourceEntry) =>
          !entry.url.includes("youtube.com") && !entry.url.includes("youtu.be")
      );
    }

    if (entries.length === 0) {
      console.log("No valid entries found in API response");
      return;
    }

    console.log(`Found ${entries.length} valid entries`);

    // Process the entries (limit to desired number)
    const processLimit = process.env.PROCESS_LIMIT
      ? parseInt(process.env.PROCESS_LIMIT)
      : 10;

    const result = await processURLs(entries, processLimit);

    console.log(
      `Added ${result.added} new entries, skipped ${result.skipped}, with ${result.errors} errors`
    );
  } catch (error) {
    console.error("Error in main function:", error);
    process.exit(1);
  } finally {
    // Close the pool when done
    await pool.end();
  }
}

// Run the script
main().catch((error) => {
  console.error("Unhandled error:", error);
  process.exit(1);
});
